File: Baekjoon/TopologicalSort/[14567]선수과목 (Prerequisite)/solution.py
N, M = list(map(int, input().split()))

parent = [[] for i in range(N+1)]
child = [0 for i in range(N+1)]
res = [0 for i in range(N+1)]


def topological_sort():
    q = []

    for i in range(1, N + 1):
        if(child[i] == 0):
            q.append((i, 1))

    while len(q) > 0:
        cv, cc = q.pop(0)

        res[cv] = cc

        for i in range(len(parent[cv])):
            child[parent[cv][i]] -= 1
            if(child[parent[cv][i]] == 0):
                q.append((parent[cv][i], cc + 1))


def print_res():
    for i in range(1, N + 1):
        print(res[i], end=" ")


if __name__ == "__main__":
    for i in range(M):
        pre, post = list(map(int, input().split()))

        parent[pre].append(post)
        child[post] += 1

    topological_sort()

    print_res()

File: Baekjoon/TopologicalSort/[14567]선수과목 (Prerequisite)/solution.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

#define endl "\n"

typedef pair<int, int> pii;

vector<vector<int>> par;
vector<int> child;
vector<int> res;
int N, M, pre, post;

void topological_sort();
void print_res();

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  cin >> N >> M;

  par = vector<vector<int>>(N + 1);
  child = vector<int>(N + 1, 0);
  res = vector<int>(N + 1, 0);

  for (int i = 0; i < M; i++) {
    cin >> pre >> post;

    par[pre].push_back(post);
    child[post]++;
  }

  topological_sort();

  print_res();

  return 0;
}

void topological_sort() {
  queue<pii> q;

  for (int i = 1; i <= N; i++) {
    if (child[i] == 0) q.push({i, 1});
  }

  while (!q.empty()) {
    int cv = q.front().first;
    int cc = q.front().second;

    q.pop();

    res[cv] = cc;

    for (int i = 0; i < par[cv].size(); i++) {
      if (--child[par[cv][i]] == 0) q.push({par[cv][i], cc + 1});
    }
  }
}

void print_res() {
  for (int i = 1; i <= N; i++)
    cout << res[i] << " ";
  cout << endl;
}